interaction-manager/src/lib_components/dropzone.js

```diff
@@ -0,0 +1,311 @@
+import actions from './actions';
+
+/**
+ * Makes an element draggable
+ */
+class Dropzone {
+    /**
+     *
+     * @param {Object} options
+     * @param {string} options.element
+     * @param {string[]} options.dropzones
+     * @param {string} options.draggedClass
+     * @param {string} options.dropzoneActiveClass
+     * @param {'switch'|'add'|'shift'|'none'} options.dropType If there is already an element in the dropzone
+     * @param {function} options.onDragStart
+     * @param {function} options.onDragMove
+     * @param {function} options.onDragEnd
+     * @param {function} options.onDropZoneEnter
+     * @param {function} options.onDropZoneLeave
+     */
+    constructor(options) {
+        const hash = (Math.random() + 1).toString(36).substring(7);
+
+        this.draggableElements = [];
+        this.dropzones = [];
+        this.draggedElement = null;
+        this.draggedOver = null;
+        this.enabled = false;
+
+        this.actionName = `drag-and-drop-${hash}`;
+        this.automaticAction = `move-to-${hash}`;
+
+        this.onMouseDown = this.onMouseDown.bind(this);
+        this.onMouseMove = this.onMouseMove.bind(this);
+        this.onMouseUp = this.onMouseUp.bind(this);
+        this.onMouseEnter = this.onMouseEnter.bind(this);
+        this.onMouseLeave = this.onMouseLeave.bind(this);
+
+        this.options = options;
+
+        this.init();
+    }
+
+    /**
+     * Get the index of the dragged item in the draggableElements
+     */
+    get draggedItemIndex() {
+        return [...this.draggableElements].indexOf(this.draggedElement);
+    }
+
+    /**
+     * Gets the body scroll offset to calculate in the dragging
+     */
+    get scrollOffset() {
+        return {
+            x: document.body.scrollLeft,
+            y: document.body.scrollTop,
+        };
+    }
+

```

Comment:
It seems  like Dropzone and Draggable share a lot of logic, does it make sense to create a base class which holds the shared code and inherit from it? It looks like we've repeated some code.

Reply:
Moved in utils/drag-base.js -> `class DragBase {` that extends `Draggable` and `Dropzone` classes.